sdk/src/utils/pdf_utils.rs

```diff
@@ -0,0 +1,507 @@
+// Copyright 2023 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License,
+// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+// or the MIT license (http://opensource.org/licenses/MIT),
+// at your option.
+
+// Unless required by applicable law or agreed to in writing,
+// this software is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
+// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
+// specific language governing permissions and limitations under
+// each license.
+
+use std::io::Write;
+
+use lopdf::{
+    dictionary, Document, Object,
+    Object::{Array, Integer, Name, Reference},
+    ObjectId, Stream,
+};
+use thiserror::Error;
+
+// Associated File Relationship
+static AF_RELATIONSHIP_KEY: &[u8] = b"AFRelationship";
+static ANNOTATIONS_KEY: &[u8] = b"Annots";
+static ASSOCIATED_FILE_KEY: &[u8] = b"AF";
+static C2PA_RELATIONSHIP: &[u8] = b"C2PA_Manifest";
+static CONTENT_CREDS: &str = "Content Credentials";
+static EMBEDDED_FILES_KEY: &[u8] = b"EmbeddedFiles";
+static SUBTYPE_KEY: &[u8] = b"Subtype";
+static NAMES_KEY: &[u8] = b"Names";
+
+#[derive(Debug, Error)]
+pub(crate) enum Error {
+    #[error(transparent)]
+    UnableToReadPdf(#[from] lopdf::Error),
+
+    #[error("Unable to add C2PA manifest as an annotation to the PDF.")]
+    AddingAnnotation,
+}
+
+const C2PA_MIME_TYPE: &str = "application/x-c2pa-manifest-store";
+
+pub(crate) trait C2paPdf: Sized {
```

Comment:
I'm curious (not at all a blocker) to see if you see the `C2paPdf` trait being used anywhere other than for the `Pdf` struct you define below. If not, what's the benefit of the abstraction?

Reply:
I added this to make unit testing easier in manifest generation code. It's hard to make PDFs that cause some functions in this trait fail. For example, `Pdf::write_manifest_as_annotation` calls `lopdf` code that can fail, but it's hard to generate a PDF that cause that to fail. Using a trait, we can instead use a mocking tool like `mockall` to do something like this: 

```rust
let mut pdf = MockC2paPdf::default();
pdf
    .expect_write_manifest_as_annotation()
    .returning(|_| Err(lopdf::Error::Blah))

// etc...
```

Happy to remove if you think it's not necessary though!

Reply:
Ahh, that makes sense. Please leave it as is.